url72c32f3c54_Quickhull3d_quickhull3d_tgz-pJ8-com_github_quickhull3d_SimpleExampleJ8/src/com/github/quickhull3d/QuickHull3D.java

```diff
@@ -0,0 +1,1334 @@
+            }
+            claimed.delete(face.outside, end);
+            end.next = null;
+            return face.outside;
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Creates an empty convex hull object.
+     */
+    public QuickHull3D() {
+    }
+
+    /**
+     * Creates a convex hull object and initializes it to the convex hull of a
+     * set of points whose coordinates are given by an array of doubles.
+     * 
+     * @param coords
+     *            x, y, and z coordinates of each input point. The length of
+     *            this array will be three times the the number of input points.
+     * @throws IllegalArgumentException
+     *             the number of input points is less than four, or the points
+     *             appear to be coincident, colinear, or coplanar.
+     */
+    public QuickHull3D(double[] coords) throws IllegalArgumentException {
+        build(coords, coords.length / 3);
+    }
+
+    /**
+     * Creates a convex hull object and initializes it to the convex hull of a
+     * set of points.
+     * 
+     * @param points
+     *            input points.
+     * @throws IllegalArgumentException
+     *             the number of input points is less than four, or the points
+     *             appear to be coincident, colinear, or coplanar.
+     */
+    public QuickHull3D(Point3d[] points) throws IllegalArgumentException {
+        build(points, points.length);
+    }
+
+    private HalfEdge findHalfEdge(Vertex tail, Vertex head) {
+        // brute force ... OK, since setHull is not used much
+        for (Iterator it = faces.iterator(); it.hasNext();) {
+            HalfEdge he = ((Face) it.next()).findEdge(tail, head);
+            if (he != null) {
+                return he;
+            }
+        }
+        return null;
+    }
+
+    protected void setHull(double[] coords, int nump, int[][] faceIndices, int numf) {
+        initBuffers(nump);
+        setPoints(coords, nump);
+        computeMaxAndMin();
+        for (int i = 0; i < numf; i++) {
+            Face face = Face.create(pointBuffer, faceIndices[i]);
+            HalfEdge he = face.he0;
+            do {
+                HalfEdge heOpp = findHalfEdge(he.head(), he.tail());
+                if (heOpp != null) {
+                    he.setOpposite(heOpp);
+                }
+                he = he.next;
+            } while (he != face.he0);
+            faces.add(face);
+        }
+    }
+
+    private void printQhullErrors(Process proc) throws IOException {
+        boolean wrote = false;
+        InputStream es = proc.getErrorStream();
+        StringBuffer error = new StringBuffer();
+        while (es.available() > 0) {
+            error.append((char) es.read());
+            wrote = true;
+        }
+        if (wrote) {
+            error.append(" ");
+            LOG.error(error.toString());
+        }
+    }
+
+    protected void setFromQhull(double[] coords, int nump, boolean triangulate) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.